File: apps/desktop/src-tauri/src/sys/commands/teams.rs
use crate::features::teams::team_manager::TeamSettings;
use crate::features::teams::{
    ActivityType, Permission, ResourceType, Team, TeamActivity, TeamActivityManager,
    TeamInvitation, TeamManager, TeamMember, TeamResource, TeamResourceManager, TeamRole,
    TeamUpdates,
};
use crate::sys::commands::auth::{get_session_user_id, SessionState};
use crate::sys::commands::AppDatabase;
use serde_json::json;
use tauri::State;

fn authorize(
    manager: &TeamManager,
    team_id: &str,
    actor_id: &str,
    permission: Permission,
) -> Result<(), String> {
    manager.authorize_permission(team_id, actor_id, permission)?;
    Ok(())
}

#[tauri::command]
pub async fn create_team(
    name: String,
    description: Option<String>,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Team, String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.create_team(name, description, actor)
}

#[tauri::command]
pub async fn get_team(
    team_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Option<Team>, String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.get_team(&team_id, &actor)
}

#[tauri::command]
pub async fn update_team(
    team_id: String,
    name: Option<String>,
    description: Option<String>,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    let updates = TeamUpdates {
        name,
        description,
        settings: None,
    };
    manager.update_team(&team_id, &actor, updates)
}

#[tauri::command]
pub async fn update_team_settings(
    team_id: String,
    default_member_role: Option<String>,
    allow_resource_sharing: Option<bool>,
    require_approval_for_automations: Option<bool>,
    enable_activity_notifications: Option<bool>,
    max_members: Option<usize>,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    let existing = manager
        .get_team(&team_id, &actor)?
        .ok_or_else(|| format!("Team not found: {}", team_id))?;

    let resolved_role = if let Some(role) = default_member_role {
        TeamRole::from_str(&role).ok_or_else(|| format!("Invalid role: {}", role))?
    } else {
        existing.settings.default_member_role
    };

    let merged_settings = TeamSettings {
        default_member_role: resolved_role,
        allow_resource_sharing: allow_resource_sharing
            .unwrap_or(existing.settings.allow_resource_sharing),
        require_approval_for_automations: require_approval_for_automations
            .unwrap_or(existing.settings.require_approval_for_automations),
        enable_activity_notifications: enable_activity_notifications
            .unwrap_or(existing.settings.enable_activity_notifications),
        max_members: max_members.or(existing.settings.max_members),
    };

    let updates = TeamUpdates {
        name: None,
        description: None,
        settings: Some(merged_settings),
    };
    manager.update_team(&team_id, &actor, updates)
}

#[tauri::command]
pub async fn delete_team(
    team_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.delete_team(&team_id, &actor)
}

#[tauri::command]
pub async fn get_user_teams(
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Vec<Team>, String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.get_user_teams(&actor)
}

#[tauri::command]
pub async fn invite_member(
    team_id: String,
    email: String,
    role: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<String, String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());

    let team_role = TeamRole::from_str(&role).ok_or_else(|| format!("Invalid role: {}", role))?;

    let invitation = manager.create_invitation(&team_id, &actor, email, team_role)?;

    let activity_manager = TeamActivityManager::new(db.conn.clone());
    activity_manager.log_activity(
        &team_id,
        Some(actor),
        ActivityType::MemberInvited,
        None,
        None,
        Some(json!({ "email": invitation.email, "role": role })),
    )?;

    Ok(invitation.token)
}

#[tauri::command]
pub async fn accept_invitation(
    token: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Team, String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    let team = manager.accept_invitation(&token, &actor)?;

    let activity_manager = TeamActivityManager::new(db.conn.clone());
    activity_manager.log_activity(
        &team.id,
        Some(actor),
        ActivityType::MemberJoined,
        None,
        None,
        None,
    )?;

    Ok(team)
}

#[tauri::command]
pub async fn remove_member(
    team_id: String,
    user_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.remove_member(&team_id, &actor, &user_id)?;

    let activity_manager = TeamActivityManager::new(db.conn.clone());
    activity_manager.log_activity(
        &team_id,
        Some(actor),
        ActivityType::MemberLeft,
        None,
        None,
        Some(json!({ "removed_user": user_id })),
    )?;

    Ok(())
}

#[tauri::command]
pub async fn update_member_role(
    team_id: String,
    user_id: String,
    role: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());

    let team_role = TeamRole::from_str(&role).ok_or_else(|| format!("Invalid role: {}", role))?;

    manager.update_member_role(&team_id, &actor, &user_id, team_role)?;

    let activity_manager = TeamActivityManager::new(db.conn.clone());
    activity_manager.log_activity(
        &team_id,
        Some(actor),
        ActivityType::MemberRoleChanged,
        None,
        None,
        Some(json!({ "user_id": user_id, "new_role": role })),
    )?;

    Ok(())
}

#[tauri::command]
pub async fn get_team_members(
    team_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Vec<TeamMember>, String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.get_team_members(&team_id, &actor)
}

#[tauri::command]
pub async fn get_team_invitations(
    team_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Vec<TeamInvitation>, String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.get_team_invitations(&team_id, &actor)
}

#[tauri::command]
pub async fn share_resource(
    team_id: String,
    resource_type: String,
    resource_id: String,
    resource_name: String,
    resource_description: Option<String>,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let res_type = ResourceType::from_str(&resource_type)
        .ok_or_else(|| format!("Invalid resource type: {}", resource_type))?;

    let teams = TeamManager::new(db.conn.clone());
    authorize(&teams, &team_id, &actor, Permission::ShareResources)?;

    let manager = TeamResourceManager::new(db.conn.clone());
    manager.share_resource(
        &team_id,
        res_type,
        &resource_id,
        resource_name.clone(),
        resource_description.clone(),
        &actor,
    )?;

    let activity_manager = TeamActivityManager::new(db.conn.clone());
    activity_manager.log_activity(
        &team_id,
        Some(actor),
        ActivityType::ResourceShared,
        Some(resource_type),
        Some(resource_id),
        Some(json!({ "name": resource_name, "description": resource_description })),
    )?;

    Ok(())
}

#[tauri::command]
pub async fn unshare_resource(
    team_id: String,
    resource_type: String,
    resource_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let res_type = ResourceType::from_str(&resource_type)
        .ok_or_else(|| format!("Invalid resource type: {}", resource_type))?;

    let teams = TeamManager::new(db.conn.clone());
    authorize(&teams, &team_id, &actor, Permission::ShareResources)?;

    let manager = TeamResourceManager::new(db.conn.clone());
    manager.unshare_resource(&team_id, res_type, &resource_id)?;

    let activity_manager = TeamActivityManager::new(db.conn.clone());
    activity_manager.log_activity(
        &team_id,
        Some(actor),
        ActivityType::ResourceUnshared,
        Some(resource_type),
        Some(resource_id),
        None,
    )?;

    Ok(())
}

#[tauri::command]
pub async fn get_team_resources(
    team_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Vec<TeamResource>, String> {
    let actor = get_session_user_id(&session)?;
    let teams = TeamManager::new(db.conn.clone());
    authorize(&teams, &team_id, &actor, Permission::ViewResources)?;

    let manager = TeamResourceManager::new(db.conn.clone());
    manager.get_team_resources(&team_id)
}

#[tauri::command]
pub async fn get_team_resources_by_type(
    team_id: String,
    resource_type: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Vec<TeamResource>, String> {
    let actor = get_session_user_id(&session)?;
    let res_type = ResourceType::from_str(&resource_type)
        .ok_or_else(|| format!("Invalid resource type: {}", resource_type))?;

    let teams = TeamManager::new(db.conn.clone());
    authorize(&teams, &team_id, &actor, Permission::ViewResources)?;

    let manager = TeamResourceManager::new(db.conn.clone());
    manager.get_team_resources_by_type(&team_id, res_type)
}

#[tauri::command]
pub async fn get_team_activity(
    team_id: String,
    limit: usize,
    offset: usize,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Vec<TeamActivity>, String> {
    let actor = get_session_user_id(&session)?;
    let teams = TeamManager::new(db.conn.clone());
    authorize(&teams, &team_id, &actor, Permission::ViewActivity)?;

    let manager = TeamActivityManager::new(db.conn.clone());
    manager.get_team_activity(&team_id, limit, offset)
}

#[tauri::command]
pub async fn get_user_team_activity(
    team_id: String,
    user_id: String,
    limit: usize,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<Vec<TeamActivity>, String> {
    let actor = get_session_user_id(&session)?;
    let teams = TeamManager::new(db.conn.clone());
    authorize(&teams, &team_id, &actor, Permission::ViewActivity)?;

    let manager = TeamActivityManager::new(db.conn.clone());
    manager.get_user_activity(&team_id, &user_id, limit)
}

#[tauri::command]
pub async fn transfer_team_ownership(
    team_id: String,
    new_owner_id: String,
    session: State<'_, SessionState>,
    db: State<'_, AppDatabase>,
) -> Result<(), String> {
    let actor = get_session_user_id(&session)?;
    let manager = TeamManager::new(db.conn.clone());
    manager.transfer_ownership(&team_id, &actor, &new_owner_id)?;

    let activity_manager = TeamActivityManager::new(db.conn.clone());
    activity_manager.log_activity(
        &team_id,
        Some(actor),
        ActivityType::MemberRoleChanged,
        None,
        None,
        Some(json!({ "new_owner": new_owner_id, "action": "ownership_transferred" })),
    )?;

    Ok(())
}

File: packages/api/src/teams.ts
/**
 * Teams API — typed wrappers for team management, members, resources, billing, and subscription commands.
 */

import { command } from '@agiworkforce/runtime';

// ---- Types ----

export interface Team {
  id: string;
  name: string;
  description?: string;
  ownerId: string;
  createdAt: string;
  memberCount: number;
}
export interface TeamMember {
  userId: string;
  role: string;
  joinedAt: string;
  email?: string;
}
export interface TeamInvitation {
  id: string;
  email: string;
  role: string;
  status: string;
  invitedBy: string;
  createdAt: string;
}
export interface TeamResource {
  resourceType: string;
  resourceId: string;
  resourceName: string;
  resourceDescription?: string;
  sharedBy: string;
  sharedAt: string;
}
export interface TeamActivity {
  id: string;
  userId: string;
  action: string;
  details?: string;
  timestamp: string;
}
export interface TeamBilling {
  teamId: string;
  plan: string;
  cycle: string;
  seatCount: number;
  costPerSeat: number;
}
export interface UsageMetrics {
  [key: string]: number;
}
export interface SubscriptionInfo {
  id: string;
  planId: string;
  status: string;
  currentPeriodEnd: string;
}
export interface PricingPlan {
  id: string;
  name: string;
  price: number;
  interval: string;
  features: string[];
}

// ---- Team CRUD ----

export async function createTeam(
  name: string,
  ownerId: string,
  description?: string,
): Promise<Team> {
  return command<Team>('create_team', { name, description, ownerId });
}
export async function getTeam(teamId: string): Promise<Team | null> {
  return command<Team | null>('get_team', { teamId });
}
export async function updateTeam(
  teamId: string,
  name?: string,
  description?: string,
): Promise<void> {
  return command<void>('update_team', { teamId, name, description });
}
export async function updateTeamSettings(
  teamId: string,
  opts: {
    defaultMemberRole?: string;
    allowResourceSharing?: boolean;
    requireApprovalForAutomations?: boolean;
    enableActivityNotifications?: boolean;
    maxMembers?: number;
  },
): Promise<void> {
  return command<void>('update_team_settings', { teamId, ...opts });
}
export async function deleteTeam(teamId: string): Promise<void> {
  return command<void>('delete_team', { teamId });
}
export async function getUserTeams(userId: string): Promise<Team[]> {
  return command<Team[]>('get_user_teams', { userId });
}
export async function transferTeamOwnership(
  teamId: string,
  newOwnerId: string,
  transferredBy: string,
): Promise<void> {
  return command<void>('transfer_team_ownership', { teamId, newOwnerId, transferredBy });
}

// ---- Members ----

export async function inviteMember(
  teamId: string,
  email: string,
  role: string,
  invitedBy: string,
): Promise<string> {
  return command<string>('invite_member', { teamId, email, role, invitedBy });
}
export async function acceptInvitation(token: string, userId: string): Promise<Team> {
  return command<Team>('accept_invitation', { token, userId });
}
export async function removeMember(
  teamId: string,
  userId: string,
  removedBy: string,
): Promise<void> {
  return command<void>('remove_member', { teamId, userId, removedBy });
}
export async function updateMemberRole(
  teamId: string,
  userId: string,
  role: string,
  updatedBy: string,
): Promise<void> {
  return command<void>('update_member_role', { teamId, userId, role, updatedBy });
}
export async function getTeamMembers(teamId: string): Promise<TeamMember[]> {
  return command<TeamMember[]>('get_team_members', { teamId });
}
export async function getTeamInvitations(teamId: string): Promise<TeamInvitation[]> {
  return command<TeamInvitation[]>('get_team_invitations', { teamId });
}

// ---- Resources ----

export async function shareResource(
  teamId: string,
  resourceType: string,
  resourceId: string,
  resourceName: string,
  sharedBy: string,
  resourceDescription?: string,
): Promise<void> {
  return command<void>('share_resource', {
    teamId,
    resourceType,
    resourceId,
    resourceName,
    resourceDescription,
    sharedBy,
  });
}
export async function unshareResource(
  teamId: string,
  resourceType: string,
  resourceId: string,
  unsharedBy: string,
): Promise<void> {
  return command<void>('unshare_resource', { teamId, resourceType, resourceId, unsharedBy });
}
export async function getTeamResources(teamId: string): Promise<TeamResource[]> {
  return command<TeamResource[]>('get_team_resources', { teamId });
}
export async function getTeamResourcesByType(
  teamId: string,
  resourceType: string,
): Promise<TeamResource[]> {
  return command<TeamResource[]>('get_team_resources_by_type', { teamId, resourceType });
}

// ---- Activity ----

export async function getTeamActivity(
  teamId: string,
  limit: number,
  offset: number,
): Promise<TeamActivity[]> {
  return command<TeamActivity[]>('get_team_activity', { teamId, limit, offset });
}
export async function getUserTeamActivity(
  teamId: string,
  userId: string,
  limit: number,
): Promise<TeamActivity[]> {
  return command<TeamActivity[]>('get_user_team_activity', { teamId, userId, limit });
}

// ---- Team Billing ----

export async function getTeamBilling(teamId: string): Promise<TeamBilling | null> {
  return command<TeamBilling | null>('get_team_billing', { teamId });
}
export async function initializeTeamBilling(
  teamId: string,
  plan: string,
  cycle: string,
  seatCount: number,
): Promise<TeamBilling> {
  return command<TeamBilling>('initialize_team_billing', { teamId, plan, cycle, seatCount });
}
export async function updateTeamPlan(
  teamId: string,
  plan: string,
  updatedBy: string,
): Promise<void> {
  return command<void>('update_team_plan', { teamId, plan, updatedBy });
}
export async function addTeamSeats(
  teamId: string,
  count: number,
  updatedBy: string,
): Promise<void> {
  return command<void>('add_team_seats', { teamId, count, updatedBy });
}
export async function removeTeamSeats(
  teamId: string,
  count: number,
  updatedBy: string,
): Promise<void> {
  return command<void>('remove_team_seats', { teamId, count, updatedBy });
}
export async function calculateTeamCost(teamId: string): Promise<number> {
  return command<number>('calculate_team_cost', { teamId });
}
export async function updateTeamUsage(teamId: string, metrics: UsageMetrics): Promise<void> {
  return command<void>('update_team_usage', { teamId, metrics });
}

// ---- Subscriptions ----

export async function subscribeToPlan(
  userId: string,
  planId: string,
  billingInterval?: string,
): Promise<SubscriptionInfo> {
  return command<SubscriptionInfo>('subscribe_to_plan', { userId, planId, billingInterval });
}
export async function upgradePlan(userId: string, newPlanId: string): Promise<SubscriptionInfo> {
  return command<SubscriptionInfo>('upgrade_plan', { userId, newPlanId });
}
export async function cancelSubscription(userId: string, subscriptionId: string): Promise<void> {
  return command<void>('cancel_subscription', { userId, subscriptionId });
}
export async function getPricingPlans(): Promise<PricingPlan[]> {
  return command<PricingPlan[]>('get_pricing_plans');
}
export async function getCurrentPlan(userId: string): Promise<PricingPlan> {
  return command<PricingPlan>('get_current_plan', { userId });
}
